scylla-cdc/src/stream_reader.rs

```diff
@@ -134,17 +139,44 @@ impl StreamReader {
             values[begin_index] = CqlValue::Timestamp(window_begin);
             values[end_index] = CqlValue::Timestamp(window_end);
 
-            let mut rows_stream = self
-                .session
-                .execute_iter(query_base.clone(), &values)
-                .await?;
-
-            let schema = CDCRowSchema::new(rows_stream.get_column_specs());
-
-            while let Some(row) = rows_stream.next().await {
-                consumer
-                    .consume_cdc(CDCRow::from_row(row?, &schema))
-                    .await?;
+            let mut sleep_after_timeout = BASIC_TIMEOUT_SLEEP;
+
+            let mut next_state = None;
+            'try_query: loop {
+                let state_clone = next_state.clone();
+                let query_res = self
+                    .session
+                    .execute_paged(&query_base, &values, next_state)
+                    .await;
+                match query_res {
+                    Ok(x) => {
+                        sleep_after_timeout = BASIC_TIMEOUT_SLEEP;
+                        if let Some(rows) = x.rows {
+                            let schema = CDCRowSchema::new(&x.col_specs);
+
+                            for row in rows {
+                                consumer
+                                    .consume_cdc(CDCRow::from_row(row, &schema))
+                                    .await?;
+                            }
+                        }
+                        match x.paging_state {
+                            Some(state) => next_state = Some(state),
+                            None => break 'try_query,
+                        }
+                    },
+                    Err(QueryError::TimeoutError) => {
+                        sleep(time::Duration::from_millis(sleep_after_timeout)).await;
+                        sleep_after_timeout = sleep_after_timeout * TIMEOUT_FACTOR;
+                        if sleep_after_timeout >= TIMEOUT_LIMIT {
+                            println!("WARN: The database keeps returning a timeout error. Trying again...");
```

Comment:
Don't use `println!` to tell things to the user, use the `tracing` crate instead. This will let the users filter out messages as they please.

Moreover, I don't think that this message is very helpful. Let's put more information here. Instead, please print a message every time there was a timeout, and include information about which streams were queried, what is the current backoff time, what was the current query window (`window_begin..window_end`) and the page number.

Reply:
I wanted to to it the same way as [here](https://github.com/piodul/scylla-cdc-rust/blob/ae8ccb5f8d03b1177b55c89c2b91dbf4176cb827/scylla-cdc/src/log_reader.rs#L180). Should it be changed, too?

Reply:
Yes, but in a separate PR.